File: 18-quick-sort/quick_sort.cpp
#include <iostream>
#include <fstream>
#include <vector>

int partition(std::vector<int>& arr, int low, int high) {
    // Choose rightmost element as pivot
    int pivot = arr[high];
    
    // Index of smaller element (indicates right position of pivot)
    int i = low - 1;
    
    for (int j = low; j < high; j++) {
        // If current element is smaller than or equal to pivot
        if (arr[j] <= pivot) {
            i++;
            std::swap(arr[i], arr[j]);
        }
    }
    std::swap(arr[i + 1], arr[high]);
    return i + 1;
}

void quickSort(std::vector<int>& arr, int low, int high) {
    if (low < high) {
        // Partitioning index
        int pi = partition(arr, low, high);
        
        // Separately sort elements before and after partition
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}

int main() {
    // Read input
    std::ifstream input("input.txt");
    int n;
    input >> n;
    
    std::vector<int> arr(n);
    for (int i = 0; i < n; i++) {
        input >> arr[i];
    }
    input.close();
    
    // Sort
    quickSort(arr, 0, n - 1);
    
    // Write output
    std::ofstream output("output.txt");
    for (int i = 0; i < n; i++) {
        output << arr[i];
        if (i < n - 1) output << " ";
    }
    output.close();
    
    return 0;
}

File: 18-quick-sort/quick_sort.py
def partition(arr: list[int], low: int, high: int) -> int:
    # Choose rightmost element as pivot
    pivot = arr[high]
    
    # Index of smaller element (indicates right position of pivot)
    i = low - 1
    
    for j in range(low, high):
        # If current element is smaller than or equal to pivot
        if arr[j] <= pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
    
    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1

def quick_sort(arr: list[int], low: int, high: int):
    if low < high:
        # Partitioning index
        pi = partition(arr, low, high)
        
        # Separately sort elements before and after partition
        quick_sort(arr, low, pi - 1)
        quick_sort(arr, pi + 1, high)

def main():
    # Read input
    with open('input.txt', 'r') as f:
        lines = f.read().strip().split('\n')
        n = int(lines[0])
        arr = list(map(int, lines[1].split()))
    
    # Sort
    quick_sort(arr, 0, n - 1)
    
    # Write output
    with open('output.txt', 'w') as f:
        f.write(' '.join(map(str, arr)))

if __name__ == "__main__":
    main()